### DISCUSSION TITLE
关于 CTRL-C

### DISCUSSION THREAD

**Discussion by q962** [xmake-io/xmake#2893] [General]
xmake r -d 

当程序运行时，能够捕获 ctrl-c 使程序中断，但是此时再次 CTRL-C 时 xmake 会退出。
并且子进程 gdb 会变成独立进程而存在（尽管在shell只输入几次回车后会退出）。

能否捕获 ctrl-c 传递给子进程？

**Comment by waruqi** [comment]
没能复现，所有子进程默认都是在一个进程组的，按理不需要捕获，所有子进程都是会收到 ctrl+c 自动退出，只有父子进程不在一个进程组，才需要捕获，还得遍历整颗进程树去挨个 kill。。

目前只有主动正常退出时候，子子进程还无法正常跟随退出，暂时没法获取到整颗进程树，我只 kill 了单层。

```console
$ xmake r -d
(lldb) target create "/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx"
Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
(lldb)
Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
(lldb)
Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
(lldb) ^C
(lldb)
ruki:xxxx ruki$ ps aux | grep lldb
ruki             82953   0.0  0.0  4277640    764 s022  S+   11:06下午   0:00.00 grep lldb
ruki:xxxx ruki$
```

**Comment by q962** [comment]
```
$ gdb
GNU gdb (Uos 8.2.1.1-1+security) 8.2.1
...

For help, type "help".
Type "apropos word" to search for commands related to "word".
>>> Quit
>>> Quit
>>> Quit
>>>

====

$ lldb
(lldb) ^C
(lldb) ^C
(lldb) ^C
(lldb) ^C
(lldb) ^C
(lldb)
```

```
$ xmake r -d pro1
(lldb) target create "/home/demo/project/demo/pro1/build/linux/x86_64/debug/pro1"
Current executable set to '/home/demo/project/demo/pro1/build/linux/x86_64/debug/pro1' (x86_64).
(lldb) ^C
(lldb)

# demo @ demo-PC in ~/project/demo/pro1 [23:15:57] C:130
$
```

```
$ xmake r -d pro1
GNU gdb (Uos 8.2.1.1-1+security) 8.2.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
BFD: warning: /home/demo/project/demo/pro1/build/linux/x86_64/release/pro1: unsupported GNU_PROPERTY_TYPE (5) type: 0xc0008002
Reading symbols from /home/demo/project/demo/pro1/build/linux/x86_64/release/pro1...(no debugging symbols found)...done.
>>>

# demo @ demo-PC in ~/project/demo/pro1 [23:17:33] C:130
$ Quit
>>>

# demo @ demo-PC in ~/project/demo/pro1 [23:17:35] C:130
$
```

关于 ctrl-c 未关闭gdb，最极端的情况是，gdb 和 shell 互相争夺输入

**Comment by waruqi** [comment]
看了下，这似乎跟 xmake 也没啥关系，是 gdb 自己捕获 ctrl-c 屏蔽了，ctrl-c 压根退不出 gdb ，只有按 q 才能退出。。

纯 gdb 运行 ctrl-c 都是退不出的，仅仅显示了 Quit 字样而已。

而 `xmake r -d` 下按 ctrl-c 原本就会因为同一个进程组，gdb 也会收到 ctrl-c 信号，从你的 log 中还能显示 Quit 字样就能看出，但是收到后 却屏蔽了它，自己并没有退出。

但是 xmake 却被 ctrl-c 退出了，所以 gdb 当然就一直存在了。。

而 lldb 不会捕获 ctrl-c 屏蔽，所以我这可以完全退干净，没遇到问题。

这种，xmake 也改不了什么，也不是 xmake 的问题。

```console
$ gdb
GNU gdb (Ubuntu 12.1-3ubuntu1) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) q
```

**Reply by q962** [reply]
之所以报告这个问题，是因为我在 lua中测试过 `os.execute("gdb")`，无论是在lua文件，还是交互模式下。
都与在终端中使用是一样的。能够捕获 CTRL-C

但是在 xmake 中就不可以，xmake 没有 os.execute 只有 os.exec

为什么会有这样的问题，只有研究过 os.execute 与 os.exec 的区别才能知道。

**Reply by waruqi** [reply]
os.execute 执行的 system() 调用 sh 命令去 wrap 的，不一样，也许 sh 还会去强杀所有子进程

https://man7.org/linux/man-pages/man3/system.3.html

>            execl("/bin/sh", "sh", "-c", command, (char *) NULL);

反正目前看 logs ，gdb 就是已经捕获到了 ctrl-c 只是自己屏蔽没去退出而已。

```
# demo @ demo-PC in ~/project/demo/pro1 [23:17:33] C:130
$ Quit
>>>
```

可以看到有 Quit ，说明它是捕获到的。

而 os.exec 直接执行的 fork/posix_spawnp，xmake 目前不会捕获 ctrl-c 内部强杀所有子进程，只有在正常退出时候，kill 单层子进程。

这个暂时改不了，xmake 不会去用 sh wrap ，都是走的原生 fork/posix_spawnp 执行，win 也是用的 CreateProcess 而不是 cmd 带起来。

或者你可以自己搞个 sh 去 wrap 下 gdb

**Comment by waruqi** [comment]
估计要设置一个 cancel_handler 去 kill 才行，system() 里面会去 kill all，跟 ctrl-c 没啥关系。。

```
/* We have to and actually can handle cancelable system().  The big
   problem: we have to kill the child process if necessary.  To do
   this a cleanup handler has to be registered and is has to be able
   to find the PID of the child.  The main problem is to reliable have
   the PID when needed.  It is not necessary for the parent thread to
   return.  It might still be in the kernel when the cancellation
   request comes.  Therefore we have to use the clone() calls ability
   to have the kernel write the PID into the user-level variable.  */
```

https://elixir.bootlin.com/uclibc-ng/v1.0.42/source/libc/stdlib/system.c
https://github.com/kraj/uClibc/blob/master/libc/stdlib/system.c
https://man7.org/tlpi/code/online/dist/procexec/system.c.html
https://github.com/lattera/glibc/blob/master/sysdeps/posix/system.c

**Comment by q962** [comment]
gdb 输出 quit，因为父进程退出了，gdb 有理由也退出进程（但是不知道为什么会存在延时）

解决方案分两步：
1. catch sigint 
2. waitpid( debugger_pid )

你可以试试
https://gist.github.com/q962/a8359541f9ad5f6b2b0a74e5ba738669

**Reply by waruqi** [reply]
> gdb 输出 quit，因为父进程退出了，gdb 有理由也退出进程（但是不知道为什么会存在延时）
> 
> 解决方案分两步：
> 
> 1. catch sigint
> 2. waitpid( debugger_pid )
> 
> 你可以试试 https://gist.github.com/q962/a8359541f9ad5f6b2b0a74e5ba738669

这两跟这个完全没关系，waitpid 只是等待进程退出，这原本就是需要的。它这里捕获 sigint 也没做任何处理，而且 gdb 原本就能收到 sigint 之前说过的

**Reply by waruqi** [reply]
> gdb 有理由也退出进程（但是不知道为什么会存在延时）

不是延时，是 gdb 收到 ctrl-c 压根不会退，敲多少次都是，即使直接运行的

**Comment by q962** [comment]
父进程接收到 sigint 没有捕获，默认操作是退出进程，gdb 也会受到 sigint捕获了，但是他可以判断父进程时候存在，不存在就退出进程。
在某个环节存在问题，导致延时在退出。

如果gdb 收到 sigint后能够及时判断父进程不存在那应该及时退出。

代码片段已经给你了，为什么不试试。我反正测试过了

```
$ ./build/linux/x86_64/release/catch_signal
self pid: 4026599
GNU gdb (Uos 8.2.1.1-1+security) 8.2.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
>>> Quit
Caught SIGINT
>>> QuitCaught SIGINT

>>> Caught SIGINT
Quit
>>> QuitCaught SIGINT

>>>
```

**Reply by waruqi** [reply]
> 父进程接收到 sigint 没有捕获，默认操作是退出进程，gdb 也会受到 sigint捕获了，但是他可以判断父进程时候存在，不存在就退出进程。

但是现在 gdb 即使直接运行，没父进程，它也没退出。。

```console
$ gdb
For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
(gdb) Quit
```

> 他可以判断父进程时候存在，不存在就退出进程
> 如果gdb 收到 sigint后能够及时判断父进程不存在那应该及时退出。

没明白，gdb应该干什么，能够干什么不是我能控制的，gdb 收到 sigint 的处理逻辑，是 gdb 内部代码逻辑的事情，我没法控制 gdb 内部的 signal handler 处理，除非 hook

只有调用 gdb 的进程，可以注册捕获 sigint 做额外的处理，去 kill gdb ，但是我无法干扰 gdb 内部的 sigint 处理。

我觉得你是想表达：我们自己的父进程捕获收到 sigint 后，应该判断 gdb 能否退出，然后去 kill 它？

> 代码片段已经给你了，为什么不试试。我反正测试过了

我很忙，没有太多时间，只能手机上抽空看看，刚稍微用电脑跑了下 你那个 example，也无法退出 gdb，所以不知道你想表达什么。

```console
$ ./build/linux/x86_64/release/testg 
self pid: 55703
GNU gdb (Ubuntu 12.1-3ubuntu1) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) Caught SIGINT
Quit
(gdb) 
```
我看里面仅仅只是注册 sigint ，没做任何处理，如果你是想表达注册完 sigint handler 里面，自己去 kill gdb 进程 pid ，这种方式我当然知道，当初就考虑过这种方式，但还有很多细节问题 没能很好解决，所以之前我没搞，在没有完美方案前，我暂时不会去搞。

1. 捕获 sigint handler 里面去 kill 子进程，只能 kill 掉单层子进程，子子进程 kill 不掉(gdb或者其他子进程起得子进程)，不完美。
2. 如果啥也不干，都在同一进程组，gdb 原本就能收到 sigint，但是它没能退出，所以只能借助 kill，在父进程 kill 
3. 我之前想到的方案是，所有子进程在同一个单独进程组，然后捕获 sigint 后，走 kill 全部干掉，但是这样实现逻辑还是很麻烦，所以之前懒得搞

**Comment by q962** [comment]
十分抱歉，是我的描述错误。

> 能否捕获 ctrl-c 传递给子进程?

知识点不足导致严重的描述错误。

我不是想在收到到sigint 后信号杀死gdb。

而是想收到 sigint 信号后xmake任然存在，方便调试。


十分抱歉😵

**Reply by waruqi** [reply]
这就没必要了吧，ctrl-c 不想让 xmake 退出，也不想让 gdb 退出？那你不敲 ctrl-c 不就好了。。 

没明白这么做如何方便调试了，如果让 gdb 退出，也可以按 q ，也不需要 ctrl-c

只是为了个人习惯，经常敲 ctrl-c ，然后又不想它退出？

**Reply by q962** [reply]
并不是因为个人习惯，gdb lldb 被设计为屏蔽sigint，但是通过 xmake 却直接退出了。

这不是和使用shell 是一样的吗？

上一秒想要执行的指令，不想执行了。
不支持 ctrl-c 那只能退格键按住，删除所有。
或者 ctrl-a 然后再输入注释符号在回车。多麻烦。

要是手误多按了一次，已经设置好的配置就丢失了。又得从来一遍。

**Reply by waruqi** [reply]
> 并不是因为个人习惯，gdb lldb 被设计为屏蔽sigint

也就gdb 这样吧，lldb 我这一 ctrl-c 就直接退了

**Reply by q962** [reply]
$ lldb-13
(lldb) ^C
(lldb) ^C
(lldb) ^C
(lldb) ^C
(lldb)

那我的咋不退

**Reply by waruqi** [reply]
> 没能复现，所有子进程默认都是在一个进程组的，按理不需要捕获，所有子进程都是会收到 ctrl+c 自动退出，只有父子进程不在一个进程组，才需要捕获，还得遍历整颗进程树去挨个 kill。。
> 
> 目前只有主动正常退出时候，子子进程还无法正常跟随退出，暂时没法获取到整颗进程树，我只 kill 了单层。
> 
> ```
> $ xmake r -d
> (lldb) target create "/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx"
> Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
> (lldb)
> Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
> (lldb)
> Current executable set to '/private/tmp/xxxx/build/macosx/x86_64/debug/xxxx' (x86_64).
> (lldb) ^C
> (lldb)
> ruki:xxxx ruki$ ps aux | grep lldb
> ruki             82953   0.0  0.0  4277640    764 s022  S+   11:06下午   0:00.00 grep lldb
> ruki:xxxx ruki$
> ```

我上面贴过，xmake 带起 lldb 去 crtl-c 都直接退干净了，现在我没电脑 没法继续验证，要等后面几天了

**Reply by q962** [reply]
> 也就gdb 这样吧，lldb 我这一 ctrl-c 就直接退了

是通过 shell 执行的 lldb，还是 `xmake r -d`

如果是 shell 那我就不知道了。在我的印象中都是不退的。

如果是 `xmake r -d` lldb 退出是正常的。因为 xmake 没有捕获 SIGINT

```
$ ./build/linux/x86_64/release/catch_signal -p lldb
p: lldb
self pid: 4082681
(lldb) target create "gdb"
Current executable set to 'gdb' (x86_64).
(lldb) ^C
Caught SIGINT
(lldb) ^C
Caught SIGINT
(lldb) ^D
Caught SIGCHLD
pid: 4082682

#######################################################

# demo @ demo-PC in ~/project/demo/catch_signal [20:49:03] C:130
$ ./build/linux/x86_64/release/catch_signal -p lldb -C # -C 不捕捉信号
p: lldb
C: (null)
self pid: 4083164
(lldb) target create "gdb"
Current executable set to 'gdb' (x86_64).
(lldb)
^C

# demo @ demo-PC in ~/project/demo/catch_signal [20:49:04] C:130

######################################################

# demo @ demo-PC in ~/project/demo/catch_signal [20:50:37]
$ xmake r -d
(lldb) target create "/home/demo/project/demo/catch_signal/build/linux/x86_64/debugger=lldb/catch_signal"
Current executable set to '/home/demo/project/demo/catch_signal/build/linux/x86_64/debugger=lldb/catch_signal' (x86_64).
(lldb) ^C
(lldb)

# demo @ demo-PC in ~/project/demo/catch_signal [20:50:42] C:130
```

你看，不捕获信号是不是和 `xmake r -d` 是一样的效果。

小问题，也不急着解决。

假期愉快

**Reply by waruqi** [reply]
看了下，要在执行进程时候，这里做一些 SIGINT/SIGQUIT 处理才行，https://github.com/kraj/uClibc/blob/ca1c74d67dd115d059a875150e10b8560a9c35a8/libc/stdlib/system.c#L38-L46

等回头有空可以考虑改进下。

**Reply by waruqi** [reply]
更新到 dev 再试试，应该可以了，xmake update dev
